### PR TITLE
Sketch for dcp-community README and charter process

### PR DESCRIPTION
**Pull request by brianraymor**
There are a number of teams interested in writing charters for existing projects or proposing new projects. This is an early draft to unblock experimentation and receive feedback on the process and template. I'm especially curious if the notification workflow can be simplified to rely only on github queries and notifications rather than slack. The lack of forums or mailing lists limit our communication options at the moment. It may also be helpful to create a tech-architecture github team to be assigned as the reviewer.

**Open issues**:
* [The value of milestones in charters](https://github.com/HumanCellAtlas/dcp-community/pull/1#discussion_r203871021)
* [Non-technical or process charters](https://github.com/HumanCellAtlas/dcp-community/pull/1#discussion_r203919084)

**Bikesheds**:
* [Product owners or PMs?](https://github.com/HumanCellAtlas/dcp-community/pull/1#discussion_r204457066)
* [Product, Project, Module, Component?](https://github.com/HumanCellAtlas/dcp-community/pull/1#discussion_r204456321)

**Pending non-editorial changes:**
* Add Science PM review to ensure _alignment with HCA Governance_ to charter process
* Add `Scientific guardrails` section to charter template

### CODE REVIEW

File: charters/charter-template.md
Comment (brianraymor) [inline charters/charter-template.md:1]: Once there's agreement on the template outline, I will add detailed descriptions for each area.

Comment (jodihir) [reply charters/charter-template.md:1]: Looks good to me - my only comment is that some non-technical projects might benefit from a charter too, so we might want to "genericize" the last couple of steps here. For example, if we decide to create a DCP communications team it may be helpful to write up a charter for it.

Comment (TimothyTickle) [reply charters/charter-template.md:1]: This process is slightly different then what was presented in the Appendix of the PM Retrospective presentation. https://docs.google.com/presentation/d/1c2pfeck-b79Pjpo0-_E-2pKnePdX38XoDnezOryEFCo/edit#slide=id.g3c820c8e83_0_58 . I am reading this as the technical architecture team always approves all charters for the HCA DCP. This is different in that the Appendix where the PM++ or technical architecture team could approve depending on the nature of the charter. If I read this correctly, what motivates this change?

Also, is it the responsibility of the approver to determine if the project is against HCA governance? From the intro of your slides, I can appreciate the want to focus on processes and decisions within the DCP scope but that creates an incomplete system in my mind that is detached from our first purpose, to make the DCP for the HCA project.

File: charters/charter-template.md
Comment (TimothyTickle) [inline charters/charter-template.md]: Can you talk a little more about milestones and why they would be in a charter? I can see a general definition of functionality that describes the new process but the milestones would change and grow and seem to be something that would be addressed iteratively in other documents.

File: charters/charter-template.md
Comment (TimothyTickle) [inline charters/charter-template.md:30]: So this is interesting because it formalizes modes of communication. I agree that we need formal channels of communication but wonder if "slack", "mailing lists", and "forums" will always be the case. I am ok with adding under ## Communication a line like "(eg. slack channel(s), mailing list(s), and discussion forum(s))" but making them sections in the template translates to me as formalizes them which I am not sure is the intention.

Comment (brianraymor) [reply charters/charter-template.md:30]: Absolutely correct. It was _shorthand_ on my part because I did not include descriptions for template sections in the sketch. The intent is to capture information to populate the community page as projects are chartered. I'll update in the next commit.

File: charters/charter-template.md
Comment (brianraymor) [reply charters/charter-template.md]: Milestones for software projects are recommended for oversight. A charter is a contract with the HCA DCP community, which commits the project owner to meet explicit high-level milestones and deliver specific “products”.  And once the milestones are achieved, the project owner can either renegotiate the charter or conclude. If there is a failure to deliver, then the charter and ownership can be reviewed. 

For governance charters such as PM++ and Architecture++, there would be no milestones.

File: charters/charter-template.md
Comment (brianraymor) [reply charters/charter-template.md:1]: @jodihir and @TimothyTickle - regarding your comments on charters for _non-technical projects_ , I intentionally did not include a path for governance/process charters to be approved by DCP PM, as originally proposed in Appendix A of the PM Retrospective. The change was motivated by conversations at the meetings where strong preferences were shared to keep the charter process simple and focused on software/technical charters.

 

The original sketch was an experiment to see if there was a clear demand for such charters with specific plans to write ​(more than?) ​one. It’s simple for me to add a “process” charter to the workflow with approval by DCP PM, but I would prefer the change to be prompted by requirements rather than “completeness”.

Comment (brianraymor) [reply charters/charter-template.md:1]: @TimothyTickle - regarding _is it the responsibility of the approver to determine if the project is against HCA governance?_ - could you offer a realistic and concrete example to help my understanding? In general, I feel like the DCP community would never allow such a proposed charter to demonstrate consensus and reach Architecture for approval.

File: README.md
Comment (brianraymor) [inline README.md]: I expect that the *Communicating* section will be replaced with a link to datahumancellatlas.org which will become the _community entrance_ to HCA DCP.

File: charters/charter-template.md
Comment (TimothyTickle) [reply charters/charter-template.md]: I can see how they may be useful for shorter-term and mid-term projects. What about long-term projects like the blue/green/purple infrastructure?

File: charters/charter-template.md
Comment (TimothyTickle) [reply charters/charter-template.md:30]: Thank you, that resolves this comment from my perspective.

File: README.md
Comment (TimothyTickle) [reply README.md]: Yes and then also maybe specifically the about us and contact us sections.

File: charters/charter-template.md
Comment (TimothyTickle) [reply charters/charter-template.md:1]: **Charters for non-technical projects**
I am ok with that, we can wait to when we need it. That being said, it should be clear that these are technical charters; what it means to be a techinical charter and what kinds of things this covers. And I think that definition should in some way be colocated with the technical charter template.

Comment (TimothyTickle) [reply charters/charter-template.md:1]: **HCA Governance Alignment**
I think it has not happened yet but has almost happened several times. It often arises when user requests are not fully aligned to HCA governance and we are split between the two. I am just struggling to understand how this is addressed not implicitly but with a specific point in the process.

Comment (brianraymor) [reply charters/charter-template.md:1]: @TimothyTickle - I can understand HCA Governance concerns for rfcs, but less so for charters which is why I was curious if there was a specific example in mind. My expectation is that such cases would be immediately flagged during community review - for example, by you. We could add a statement that _Science PM_ (Jonah) would review for alignment with HCA Science governance. [Discussed with @TimothyTickle - will add this step to the process]

File: charters/charter-template.md
Comment (brianraymor) [reply charters/charter-template.md]: Most standards efforts last 2-3 years which I consider long-term. I would recommend chartering for shorter periods and re-chartering based on updated scope. OASIS requires a re-charter if the scope of a technical committee changes.  You might update the Secondary Analysis charter when new pipelines are approved by HCA Science governance. 

If you strongly disagree, I would recommend reviewing with the other Leads to see if there is a consensus position among you.

Comment (kbergin) [reply charters/charter-template.md]: I agree with Tim that listing milestones here is anti-agile. I could see writing the vision of the product and a link to a product roadmap.

File: README.md
Comment (kbergin) [inline README.md]: I'm trying to think of an example of a project in the DCP. As far as I can think of, we are all making products. Do you have an example in mind?

File: charters/README.md
Comment (kbergin) [inline charters/README.md:3]: Again, we're making products.

File: charters/charter-template.md
Comment (kbergin) [inline charters/charter-template.md]: As stated over email, @TimothyTickle and my vote here is for coordinator, product owner, and technical lead to be the roles.

File: charters/README.md
Comment (kbergin) [reply charters/README.md:3]: Assume my vote is for replacement of the word project throughout all of this PR.

Comment (brianraymor) [reply charters/README.md:3]: @kbergin - From my perspective, DCP is the _product_ not the individual boxes. How about _component_?

File: charters/charter-template.md
Comment (brianraymor) [reply charters/charter-template.md]: Tim proposed _Coordinator_ on an email thread with the current leads. At this point, the only response is a preference to maintain _Lead_ from the DCP Quarterly goals in favor of _Project Lead_.  My approach is to allow the current membership of each role (lead, product owner, and technical lead) to reach consensus on their own titles and responsibilities.

Comment (kislyuk) [reply charters/charter-template.md]: I would like us to avoid using the term "product owner" as it is ambiguous and misleading. Let's use the next best acceptable alternative.

Comment (kbergin) [reply charters/charter-template.md]: @kislyuk Do you have a suggestion on another alternative? It is a pretty defined role for someone who is owner of the product. In reality, all role names are ambiguous and depend on the institution using them and the person in them. Technical Lead and Project Lead are just as ambiguous.

If we were to have definitions of the role somewhere would that satisfy the requirement?

File: charters/README.md
Comment (kbergin) [reply charters/README.md:3]: Each box can be reused in different combinations and intend to be reusable with other products in the data biosphere. With that in mind, each one is a product. Regardless, we should be applying product thinking to the development of all of them.

Comment (mweiden) [inline charters/README.md:3]: Perhaps this should call out technical projects specifically, given that there was confusion about this.

File: charters/README.md
Comment (mweiden) [inline charters/README.md]: Consider clarifying the word "roadmap." There is a chance that first time users understand it to mean a "set of project milestones" rather than a "a set of thematic goals," which is what I think you mean.

File: charters/charter-template.md
Comment (mweiden) [inline charters/charter-template.md:19]: I don't know if adding this is in the scope of this PR, but we need very precise, public documentation on what the remit of each of these roles is.

File: charters/README.md
Comment (TimothyTickle) [reply charters/README.md:3]: I agree with @kbergin here. I think if it was a monolithic system I would go with component but given we are making a collection of reusable software functionality where each should stand on its own (biosphere style) each piece should be able to have a life of its own and should be considered a product. The DCP in my mind is a collection of integrated products, each serving a different need in a reusable way (or so we aspire ;-) ).

File: charters/charter-template.md
Comment (brianraymor) [reply charters/charter-template.md:19]: It is in scope. The topic is currently under active discussion with members from each group (lead, product owner, technical lead) to formalize role name and responsibilities.

File: charters/charter-template.md
Comment (mweiden) [inline charters/charter-template.md:9]: Re. module alignment with HCA Governance, one option is to add a `Scientific guardrails` subsection here. A `Scientific guardrails` section could describe contact people and processes that governing and science orgs. hold with DCP modules to make sure that developments in the module suit the needs of the scientific community. This section could also mandate reviews by the governing and science orgs. of developments in the module that require oversight.

/cc @TimothyTickle @brianraymor

File: charters/README.md
Comment (brianraymor) [reply charters/README.md:3]: [Biosphere](https://medium.com/@benedictpaten/a-data-biosphere-for-biomedical-research-d212bbfae95d) describes *modular component* as:

_The Data Biosphere will work best if it is designed around key modular components — each having discrete capabilities and clear rules of interaction, and each served by multiple alternative implementations. To offer a familiar analogy, modern internet technology is organized around components that play distinct roles, such as web servers, web browsers, search engines, and email clients. For each component, users have multiple choices, which can be mixed and matched._

_By a Data Environment, we mean a suite of services, assembled from the Components, that allow users to carry out a set of related tasks in an easy way._

And the [Data Store](https://github.com/DataBiosphere/cgp-data-store) is one such modular component in their repository.

Comment (kbergin) [reply charters/README.md:3]: In that excerpt nothing is defined as a product. I think whether something is a component of a data biosphere environment and whether it in itself is a product are two different things, and one could be both. 

We build products that are able to be components in a data biosphere environment.

Comment (brianraymor) [reply charters/README.md:3]: @kbergin - That's correct. The conversation moved on. Tim responded to my suggestion to use _component_ with _given we are making a collection of reusable software functionality where each should stand on its own **(biosphere style)**_. Biosphere style is _modular components_ like the Data Store. 

In your model, is Apache Spark a product?

File: charters/charter-template.md
Comment (mweiden) [reply charters/charter-template.md]: Re. agile: @kbergin it would be anti-agile if charters remain static. Comparatively, @brianraymor suggested above re-chartering if scope is updated. We can address your core concern--that modules can adapt their scope fluidly when requirements change--if we make sure that the approval process for amending charters is responsive.

Re. oversight and transparency: presenting delivery accountabilities in a clear, standardized way will allow teams to better understand those accountabilities and remain informed on changes to them. Recording milestones here accomplishes that.

File: charters/README.md
Comment (brianraymor) [reply charters/README.md]: I'll clarify in the next commit.

File: charters/README.md
Comment (TimothyTickle) [reply charters/README.md:3]: @brianraymor I would like to roll back the conversation here. I indicated at the beginning of my comment that I agreed with Kylee; I do not agree the use of component is optimal as I currently understand charters. I believe, for the majority of things in charters, they will be best viewed as products (or at least existing work we have). We have so far thought of them this way so I am not sure of why that would change.

I think the reason the biopshere blog uses the word component while I think we would use product is really a question of implementation. In the biosphere excerpt, it mentions the components are served by multiple implementations. I see components here being described as the abstraction of the functionality (the class) while the products are the instantiations or what we build (the object). In reality, the biosphere will abstractly contain the functionality data store (component) that will be fulfilled/instantiated with work on many products (HCA blue box) that share similar interfaces/functionality. Given we are making the implementations I would prefer us to be product focused.

I could imagine a world where component (not biopshere component but component as in generically "a piece") could be the subject of some charters, if it is a focused augmentation of a product we have. In a way a supplemental piece. One example of this is if one wanted to charter the addition of another cloud store backend to the HCA data store. This would not necessarily be a freestanding product (like the HCA data store) but could be described as a component (if charters are that specific, still a little hazy on the scope). I hope that better communicates my thoughts.

Comment (TimothyTickle) [reply charters/README.md:3]: +1 to using reference documents from our project's history. I want a world where we can do the same with governance.

File: charters/charter-template.md
Comment (brianraymor) [reply charters/charter-template.md]: @kislyuk - It's not ambiguous. It may be misleading. It's a defined role specific to **Scrum** and there's a [historical reason](https://www.romanpichler.com/blog/product-manager-vs-product-owner/) that it was introduced over product manager at the time. It does imply that DCP is demanding or assuming adherence to a methodology.

Comment (kislyuk) [reply charters/charter-template.md]: I think it is ambiguous and misleading.

I think the terms "project manager", "product manager", and "technical lead" are less ambiguous.

I confess I was not actually able to follow the reasoning in your link at all, @brianraymor.

File: charters/README.md
Comment (mweiden) [reply charters/README.md:3]: I've done some looking around and have two thoughts here.
1. It appears that open source foundations [generally](https://www.freebsdfoundation.org/what-we-do/projects/) [use](https://www.eclipse.org/org/foundation/) "[project](https://www.apache.org/foundation/how-it-works.html)", although a few (with seemingly more user-facing software) use "[product](https://www.mozilla.org/en-US/about/)".
2. The best differentiation I've read thus far is that an `OSS product = (a software project) + standardization + QA + packaging + branding + support`.

The question to me is if each of the box teams provide the extra elements listed in (2) for their software individually.

File: charters/charter-template.md
Comment (brianraymor) [reply charters/charter-template.md:9]: @mweiden - Reviewed your proposal with @TimothyTickle. Including a `Scientific guardrails` section in the template and requiring an explicit review by the Science PM in the process addresses the _alignment with HCA Governance_ concerns. I'll add in the next update.

File: charters/charter-template.md
Comment (lauraclarke) [reply charters/charter-template.md]: I am with @brianraymor and @mweiden I think the milestones here are big picture goals linked to user needs rather than implementation specific tasks. There don't need to be many, nor does it need to be exhaustive. 

Looking at https://www.oasis-open.org/committees/mqtt/charter.php#item-4 their milestones was about delivering a draft of the specification and maybe a pilot implementation (not sure how to interpret OASIS standard version of the MQTT protocol specification). 

For the metadata charter, I would envision our milestones to be multiplex fish imaging metadata and nuc-seq metadata as they are the two main new assays we will target in the next few months

@brianraymor Do you have a feeling for how unusual it is to update these lists more frequently than yearly. I think for metadata, it is likely other things will emerge and hopefully, those two will take much less than a whole year?

File: charters/charter-template.md
Comment (lauraclarke) [reply charters/charter-template.md:1]: @brianraymor as I said when we spoke I think limiting charters to technical issues isn't the best plan as even with good oversight there are clear non-technical components such as data wrangling or website content where having a similar set of goals, scope, and milestones defined would be useful. I appreciate some things are perhaps a bit fuzzier in these cases but not too much more. 

I am going to work with @claymfischer to write both the metadata and support charters so hopefully we can use that as a launch point for the non-technical charters

File: charters/charter-template.md
Comment (kozbo) [reply charters/charter-template.md]: If we view milestones as product vision statements and not a list of specific deliverables then I don't see them conflicting with Agile development principles. But it is confusing to call these "milestones" as those are usually used to refer to inflection points along the effort to develop something larger.   So how about calling them "Project Vision" or "Project Goals"?

File: charters/README.md
Comment (kozbo) [reply charters/README.md:3]: I think that much of the language we are defining has already been defined in industry. I agree that each module/box is a product in itself. Products are long-running and usually have lots of checkpoints along the way.  I would normally call these checkpoints releases, but the concept of release seems to be reserved for data releases only.  I think we can view these checkpoints/releases as projects. Each charter can be viewed as a high-level description of what will be in the release. So from my software development perspective I think that we have:
**Products:** Equivalent to our  current boxes, do not require a charter
**Projects:** Equivalent to the effort to create a software release on a product
**Charter:** High level descriptions of the organization constructed to deliver a set of features or capabilities. I think that we would like to have charters for our Product groups too, but putting charters at this granularity keeps us focused on what the Product needs to deliver feature-wise.
**Project Vision:** High level vision statements for what this release is trying to achieve. (avoid using milestone as that has the connotation of a more specific short term goal in the development jargon I am used to).

Comment (brianraymor) [reply charters/README.md:3]: From our [HCA Data Coordination Platform wiki](https://github.com/HumanCellAtlas/wiki/wiki):

_As currently conceived, this data coordination platform will provide four key **components**: ingestion services for submission of data; synchronized data storage across multiple clouds; standardized secondary analysis pipelines; and portals for data access, tertiary analysis, and visualization._

File: README.md
Comment (briandoconnor) [inline README.md:1]: @brianraymor do you see this repo markdown content being pulled into, say, a community sub-section of the portal under like Development or Contribute?

https://dev.data.humancellatlas.org/develop/development-guides/development-guides-overview

or

https://dev.data.humancellatlas.org/contribute/overview/overview

File: charters/README.md
Comment (briandoconnor) [inline charters/README.md:14]: @brianraymor do you want to have an RFC/Charter-focused slack channel?

File: charters/README.md
Comment (briandoconnor) [inline charters/README.md]: @brianraymor This is the "Bi-weekly DCP PM meeting" every other Thursday at 8am pacific, right?  Once this PR is merged, it would be good to update the agenda of this meeting to point out the PMs/Tech Leads need to approve charters in that meeting.  What about RFCs?  I can hold that question, though, until you have a PR to tackle that SOP.  I'm assuming the RFC process documentation will go in this repo, is that right?

File: README.md
Comment (briandoconnor) [inline README.md]: Has the RFC process been written up yet?  I see something basic here: https://github.com/HumanCellAtlas/RFCs  Will that SOP doc's home be here or in that repo?

File: charters/README.md
Comment (briandoconnor) [reply charters/README.md:3]: I like @mweiden 's definition here that a product is the software component + a bunch of other things.  When I think about the HCA data storage service I think the software component.  When I think of Blue Box (or we call it Redwood in other projects) it's the software component + deployment scripts + documentation + slack channel/community + etc... all of which are the Blue Box (Redwood) product which is used by a wide variety of projects like HCA/Data Commons/Data STAGE/AnVIL/etc.

Anyway, this is how I'm thinking about it.  I'm pretty flexible with calling things components or products personally.  But defining it somewhere would be helpful and then staying consistent.

File: charters/charter-template.md
Comment (briandoconnor) [reply charters/charter-template.md]: I think "product owner", "project manager", "product manager", and "technical lead" are not interchangeable and each have their own specific meaning (that may differ based on who you talk to).  @brianraymor I think whatever term we use here (or terms) it's best to define what's meant by the term.  Just define it so we have a shared understanding on the term as applied to this process.

**Review by briandoconnor — APPROVED** [review]
I left comments but I think this should be merged so the next PR can start filling in the various missing sections of these docs and give us a chance to try out converting our blue and orange charters to the .md format via a pull request.

File: README.md
Comment (brianraymor) [reply README.md:1]: Per the discussion on the DCP PM call, Gabby and I are working on recommendations to _rationalize_ the relationships between:

1. https://www.humancellatlas.org/data-sharing 
2. http://data.humancellatlas.org
3. https://github.com/HumanCellAtlas/wiki/wiki
4. and this PR.

I expect to delete the _Communicating_ section from README.md and focus on governance.

File: charters/README.md
Comment (brianraymor) [reply charters/README.md:14]: I don't really want to increase the number of slack channels. I was curious whether we could eliminate the slack announcement and rely on github notifications.

File: charters/README.md
Comment (brianraymor) [reply charters/README.md]: No. I'm referring to the bi-weekly _Tech Architecture_ meeting.  Technical charters are approved by the Tech Leads. If we have process charters, those charters would be approved by PM. 

The RFC process will also be in the dcp-community repo. One stop shopping.

File: README.md
Comment (brianraymor) [reply README.md]: That repo was an early experiment by @sampierson which needs to be deleted. The community repository is the central location for all HCA DCP community governance/process.

A proposal for a RFC process can be found in [Appendix A](https://docs.google.com/presentation/d/1c2pfeck-b79Pjpo0-_E-2pKnePdX38XoDnezOryEFCo/edit#slide=id.g3c69396420_0_515) of the PM Retrospective slides.

File: charters/README.md
Comment (kozbo) [reply charters/README.md:3]: I have been having some discussions and did some research and I would like to re-pitch my definition of project and product. There *are* a few key differences between commercial software and open source terminology. Virtually all open source software is referred to as projects. An open source project is often supplemented by a commercial company to make a product. Our components cannot stand alone as products. They each depend on other supporting software to become products. We depend on each other to become the DCP product.

File: charters/charter-template.md
Comment (tburdett) [reply charters/charter-template.md]: I'm in favour of milestones as key points of alignment between components. I think this is what we have so far being trying to accomplish with our quarterly goals and roadmapping, and I'd like to have a common vision for the whole DCP that can be mapped into milestones in modules. I think allowing these to be broader in scope with less specific detail provides opportunities for evolution that *encourages* agility rather than stifles it, especially coupled with a continuous review process. I'm ambivalent about whether we call them milestones or goals.

File: charters/charter-template.md
Comment (tburdett) [reply charters/charter-template.md:1]: +1 from me on charters not being purely limited to technical. I can imagine very clear scenarios where technical components will benefit from having a view over data wrangling goals in terms of scaling and new data types, for example. It seems clear to me though that the technical charters are the more pressing concern

File: charters/README.md
Comment (tburdett) [reply charters/README.md:3]: I really like @mweiden's definition here - this split makes total sense to me. I think our role here is to deliver the DCP as a single product. I agree for sure that there will be requirements to repurpose components - or even smaller parts of the codebase - for use in other scenarios (for example, JSON schema validation with added ontology plugins is a part of ingest we are already planning to reuse) but this will come with additional customisation or packaging to make it a standalone product.

**Review by tburdett — APPROVED** [review]
I have provided some comments to the ongoing discussions, but I don't think any are blockers to merge this pull request and begin adding more detailed content, I'm happy with the sketch as written

File: charters/charter-template.md
Comment (TimothyTickle) [reply charters/charter-template.md:1]: I am not sure I fully understand how we are going to leverage charters. In that light, experimenting with technical and nontechnical seems appropriate where people feel they can apply them.

File: charters/charter-template.md
Comment (TimothyTickle) [reply charters/charter-template.md]: Goals and vision for a charter makes sense to me; I prefer that over the word milestones. We already have a quarterly goals process that updates every quarter and we need roadmap (as in long-term but vague planning) processes formalized but I do not think anchoring them to charters makes sense under my current understanding of charters. That being said, if someone else has a clear vision of how charter milestones/goals/vision, long-term, and short-term planning would interact; I would love to read it here.

**Pull request merged**
